src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,790 @@
+            return (self.fq_type == other.fq_type
+                    and self.var == other.var
+                    and self.prime() == other.prime()
+                    and self.modulus() == other.modulus())
+        else:
+            raise TypeError(f"Cannot compare fq_default_ctx with {type(other)}")
+
+    def __hash__(self):
+        return hash((self.type, self.var, self.prime(), self.modulus()))
+
+    def __str__(self):
+        if self.degree() == 1:
+            return f"Context for fq_default in GF({self.prime()})"
+        return f"Context for fq_default in GF({self.prime()}^{self.degree()})[{self.var.decode()}]/({self.modulus().str(var=self.var.decode())})"
+
+    def __repr__(self):
+        if self.degree() == 1:
+            return f"fq_default_ctx({self.prime()}, var='{self.var.decode()}' type='{self.fq_type_str}')"
+        return f"fq_default_ctx({self.prime()}, {self.degree()}, '{self.var.decode()}', {self.modulus()!r}, '{self.fq_type_str}')"
+
+    def __call__(self, val):
+        return fq_default(val, self)
+
+
+cdef class fq_default(flint_scalar):
+    def __cinit__(self, val, ctx):
+        if not typecheck(ctx, fq_default_ctx):
+            raise TypeError
+        self.ctx = ctx
+        fq_default_init(self.val, self.ctx.val)
+
+    def __dealloc__(self):
+        if self.ctx is not None:
+            fq_default_clear(self.val, self.ctx.val)
+
+    def __init__(self, val, ctx):
+        if not typecheck(ctx, fq_default_ctx):
+            raise TypeError
+        self.ctx = ctx
+
+        check = self.ctx.set_any_as_fq_default(self.val, val)
+        if check is NotImplemented:
+            raise TypeError
+
+    def __int__(self):
+        """
+        Attempts to lift self to an integer of type fmpz in [0, p-1]
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> a = gf([1,2,3])
+            >>> int(gf(123))
+            123
+        """
+        cdef fmpz x = fmpz.__new__(fmpz)
+        res = fq_default_get_fmpz(x.val, self.val, self.ctx.val)
+        if res == 1:
+            return int(x)
+        raise ValueError("fq element has no lift to the integers")
+
+    def polynomial(self):
+        """
+        Returns a representative of ``self`` as a polynomial in `(Z/pZ)[x] / h(x)`
+        where `h(x)` is the defining polynomial of the finite field.
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> a = gf([1,2,3])
+            >>> a.polynomial()
+            3*x^2 + 2*x + 1
+            >>> gf(123).polynomial()
+            123
+        """
+        cdef fmpz_mod_poly pol
+
+        ring_ctx = fmpz_mod_poly_ctx(self.ctx.prime())
+        pol = ring_ctx.new_ctype_poly()
+        fq_default_get_fmpz_mod_poly((<fmpz_mod_poly>pol).val, self.val, self.ctx.val)
+
+        return pol
+
+    def str(self):
+        return self.polynomial().str(var=self.ctx.var.decode())
+
+    def __repr__(self):
+        # TODO: what do we want here?
+        return str(self)
```

Comment:
I'm not sure but maybe it is better to define `.repr()` here and have the superclass figure out which one to call based on the pretty setting. I think that is done for some other types if I remember correctly.

Reply:
I think i have done this as you want